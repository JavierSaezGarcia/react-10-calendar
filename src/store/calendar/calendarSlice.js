import { createSlice } from '@reduxjs/toolkit';
import { addHours } from 'date-fns';

const tempEvent = {
    _id: new Date().getTime(),
    title: "Cumpleaños del jefe",  // obligatorio
    notes: "necesito un presente",  // opcional
    start: new Date(),              // obligatorio
    end: addHours(new Date(), 2),   // obligatorio
    bgColor: "#347CF7",             // opcional
    user: {
      _id: "123",
      name: "Javier Sáez"
    }
  }

export const calendarSlice = createSlice({
  name: 'calendar',
  initialState: {
    events: [ tempEvent ],
    activeEvent: null
  },
  reducers: {
    onSetActiveEvent: ( state, { payload } ) => {
      state.activeEvent = payload;    
      // console.log(payload);
    },
    onAddNewEvent: ( state, { payload } ) => {
      state.events.push( payload );
      state.activeEvent = null;
    
      // console.log(payload);
    },
    onUpdateEvent: ( state, { payload } ) => {
      state.events = state.events.map( event => {
        if ( event._id === payload._id ) {
          return payload;
        }
        return event;
        
       
      });
    },
    onDeleteEvent: ( state ) => { 
      if(state.activeEvent){
        state.events = state.events.filter( event => event._id !== state.activeEvent._id );
        state.activeEvent = null;
      }
      
    }
  }
});

export const { onSetActiveEvent, onAddNewEvent, onUpdateEvent, onDeleteEvent } = calendarSlice.actions
