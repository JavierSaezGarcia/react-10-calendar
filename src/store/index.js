export * from './calendar/calendarSlice';
export * from './ui/uiSlice';

export * from './store'; // el store SIEMPRE SE EXPORTA EL ULTIMO

