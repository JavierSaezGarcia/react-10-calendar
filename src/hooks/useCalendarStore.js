import { useSelector, useDispatch } from "react-redux";
import { calendarApi } from "../api";
import { onSetActiveEvent, onAddNewEvent, onUpdateEvent, onDeleteEvent, onLoadEvents } from "../store";
import Swal from 'sweetalert2';
import { convertEventsToDateEvents } from "../helpers";


export const useCalendarStore = () => {

    const dispatch = useDispatch();

    const { events, activeEvent } = useSelector(state => state.calendar);    

    const { user } = useSelector(state => state.auth);


    const setActiveEvent = (calendarEvent) => {
        dispatch(onSetActiveEvent(calendarEvent))
    }

    const startSavingEvent = async (calendarEvent) => {
       
        try {
            if (calendarEvent.id) {
                // Actualizando
                console.log('Actualizando Event');
                await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
                dispatch(onUpdateEvent({ ...calendarEvent, user }));
                return;

            }
           
            calendarEvent.user = user.uid;
            // console.log('este es el calendar event que guardo despues:', calendarEvent);
            const { data } = await calendarApi.post('/events', calendarEvent );
            dispatch( onAddNewEvent({ ...calendarEvent, id: data.event.id, user }) );          



        } catch (error) {
            //  console.log(error);
            Swal.fire('Error al guardar', error.response.data.msg, 'error');
        }

    }

    const startDeletingEvent = () => {
        dispatch(onDeleteEvent(activeEvent._id))
    }

    const startLoadingEvents = async () => {
        try {

            const { data } = await calendarApi.get('/events');
            // console.log({ data });
            const events = convertEventsToDateEvents(data.events);
            console.log(events);
            dispatch(onLoadEvents(events));

        } catch (error) {
            console.log('Cargando eventos');
            console.log(error);
        }
    }

    return {
        //! Propiedades
        events,
        activeEvent,
        hasActiveEvent: !!activeEvent,

        //! Metodos
        startDeletingEvent,
        setActiveEvent,
        startSavingEvent,
        startLoadingEvents
    }
}
