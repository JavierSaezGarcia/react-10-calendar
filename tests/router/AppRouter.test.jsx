
// TODO testear que checkAuthToken haya sido llamado
// TODO testear qiue el estado sea checking

import { fireEvent, render, screen } from "@testing-library/react";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { AppRouter } from "../../src/router/AppRouter";
import { MemoryRouter } from "react-router-dom";
import { CalendarPage } from "../../src/calendar";


jest.mock('../../src/hooks/useAuthStore');
jest.mock('../../src/calendar', () => ({
    CalendarPage: () => <h1>CalendarPage</h1>
}));


describe('Test in <AppRouter />', () => {

    const mockCheckAuthToken = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    test('should display loading screen and checkAuthToken has been called ', async () => {

        useAuthStore.mockReturnValue({
            status: 'checking',
            checkAuthToken: mockCheckAuthToken

        });
        render(<AppRouter />);
        // screen.debug();

        expect(screen.getByText('Loading...')).toBeTruthy();
        expect(mockCheckAuthToken).toHaveBeenCalled();
    });

    test('should display login screen when user is not logged in', () => {

        useAuthStore.mockReturnValue({
            status: 'not-authenticated',
            checkAuthToken: mockCheckAuthToken
        });

        const { container } = render(

            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>

        );
        //screen.debug();

        expect(screen.getByText('Ingreso')).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    test('should display calendar screen when user is logged in', () => {
        useAuthStore.mockReturnValue({
            status: 'authenticated',
            checkAuthToken: mockCheckAuthToken
        });

        render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        );
        screen.debug();
        expect(screen.getByText('CalendarPage')).toBeTruthy();

    })




});