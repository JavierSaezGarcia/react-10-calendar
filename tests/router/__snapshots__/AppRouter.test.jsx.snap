// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AppRouter /> should display calendar screen when user is logged in 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in <AppRouter /> should display calendar when user is not logged in 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in <AppRouter /> should display login screen when user is not logged in 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in <AppRouter /> should display login screen when user is not logged in 2`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
